bpo-35047: Better error messages in unittest.mock

PetterS opened the pull request:
When developing unit tests with `unittest.mock`, it is common to see error messages like this:

    AssertionError: Expected 'info' to not have been called. Called 3 times.

This commit adds the actual list of calls to the error message.

<!-- issue-number: [bpo-34547](https://bugs.python.org/issue34547) -->
https://bugs.python.org/issue35047
<!-- /issue-number -->

PetterS committed:
bpo-35047: Better error messages in unittest.mock

When developing unit tests with `unittest.mock`, it is common to see error messages like this:

    AssertionError: Expected 'info' to not have been called. Called 3 times.

This commit adds the actual list of calls to the error message.

eamanu reviewed (approved):
it's ok for me.

vstinner commented on Lib/unittest/mock.py:
```diff
@@ -796,8 +798,10 @@ def assert_called_once(_mock_self):
         if not self.call_count == 1:
-            msg = ("Expected '%s' to have been called once. Called %s times." %
-                   (self._mock_name or 'mock', self.call_count))
+            msg = ("Expected '%s' to have been called once. Called %s times.\n"
```
The error message is different but very close. I suggest to write an helper method or function to format the exception message.

PetterS replied:
Agreed. Will do and use it to limit the output length for existing code as well.

vstinner commented on Lib/unittest/mock.py:
```diff
@@ -778,8 +778,10 @@ def assert_not_called(_mock_self):
-            msg = ("Expected '%s' to not have been called. Called %s times." %
-                   (self._mock_name or 'mock', self.call_count))
+            msg = ("Expected '%s' to not have been called. Called %s times.\n"
+                   "Calls: %r." % (self._mock_name or 'mock',
```
repr(self.mock_calls) can be very long.

unittest.utils has a safe_repr() method which truncates the repr to 80 characters. Maybe you should use it?

Alternative: there is also reprlib.repr(), but I suggest to reuse safe_repr() instead.

mariocj89 replied:
Agree, scares me out that someone mocks a heavily used functions and this prints all the calls.

PetterS replied:
I agree. But the reason I didn't was that `assert_has_calls` already has this functionality and does not limit the output. I should probably change that as well.

vstinner reviewed (changes requested):
What about other methods like assert_called_once_with()?

If you use safe_repr(), please modify other exceptions including mock_calls, like assert_has_calls().

cc @mariocj89

mariocj89 reviewed (commented):
Note: The link in the PR description does not match the one in the title.

Also, there are no tests validating this change :)

mariocj89 commented on Lib/unittest/mock.py:
```diff
@@ -796,8 +798,10 @@ def assert_called_once(_mock_self):
-            msg = ("Expected '%s' to have been called once. Called %s times." %
-                   (self._mock_name or 'mock', self.call_count))
+            msg = ("Expected '%s' to have been called once. Called %s times.\n"
+                   "Calls %r. " % (self._mock_name or 'mock',
```
If you get this into a helper method, I'd remove this when the mock is not called, as it will look weird to see this line with no calls, it would just add noise

vstinner replied:
I concur. If there is no call, just omit "Calls..." ;-)

PetterS replied:
Done.

PetterS committed:
bpo-35047: Add helper method and unit tests.

mariocj89 commented on Lib/unittest/test/testmock/testmock.py:
```diff
@@ -1271,6 +1286,20 @@ def test_assert_called_once(self):
+        m = Mock()
+        with self.assertRaises(AssertionError) as e:
+            m.assert_called_once()
+        self.assertFalse("Calls:" in str(e.exception))
```
`self.assertNotIn("Calls:", str(e.exception))` will give a nicer error message if it fails.

PetterS replied:
Thanks.

PetterS committed:
bpo-35047: Run make patchcheck.

PetterS committed:
bpo-35047: Use assertNotIn.

vstinner commented on Lib/unittest/mock.py:
```diff
@@ -847,8 +854,8 @@ def assert_has_calls(self, calls, any_order=False):
-                    'Calls not found.\nExpected: %r\n'
-                    'Actual: %r' % (_CallList(calls), self.mock_calls)
+                    'Calls not found.\nExpected: %r%s'
+                    % (_CallList(calls), self._call_list_as_string())
```
Hum, I would prefer to not replace "Actual" with "Calls". Maybe add an optional parameter to _call_list_as_string(prefix="Calls")?

vstinner commented on Lib/unittest/mock.py:
```diff
@@ -909,6 +916,19 @@ def _get_child_mock(self, **kw):
         return klass(**kw)
 
 
+    def _call_list_as_string(self):
```
I suggest to rename it to: _calls_repr().

vstinner commented on Lib/unittest/test/testmock/testmock.py:
```diff
@@ -407,6 +407,14 @@ def test_assert_called_once_with(self):
+        m(1)
+        m(2)
+        self.assertRaisesRegex(AssertionError,
+            r"Calls: \[call\(1\), call\(2\)\]",
```
You may use re.escape() instead. Or maybe use "with assertRaises(AssertionError) as cm: ...; self.assertIn("Calls: ...", str(cm.exception))".

vstinner commented on Misc/NEWS.d/next/Library/2018-10-25-09-59-00.bpo-35047.abbaa.rst:
```diff
@@ -0,0 +1,2 @@
+``unittest.mock`` now prints the list of calls if ``assert_not_called``
```
To me, "print" means writing into stdout. I would prefer to say something like "now includes mock calls in exception message".

PetterS committed:
bpo-35047: Changes per review.

- Rename helper function. Add optional argument.
- Use re.compile.
- Change NEWS entry.

vstinner reviewed (approved):
LGTM.

mariocj89 commented on Misc/NEWS.d/next/Library/2018-10-25-09-59-00.bpo-35047.abbaa.rst:
```diff
@@ -0,0 +1,3 @@
+``unittest.mock`` now includes mock calls in exception messages if
```
Nitpick: the change affects also `assert_called_once_with` and `assert_has_calls`

PetterS replied:
`assert_has_calls` was only refactored. Is it OK to write `assert_called_once(_with)`?

mariocj89 replied:
Good point!

About assert_called_once_with, I'd just list the three

PetterS replied:
Done!

PetterS committed:
bpo-35047: List all three functions in the NEWS entry.

mariocj89 reviewed (approved):
LGTM. Thanks for the patience going through the review :)
